tests/rptest/tests/compatibility/kafka_streams_test.py

```diff
@@ -446,21 +424,187 @@ def test_kafka_streams_map_function(self):
+                       lambda text: "hello" in text)
+
+        # HTTP GET on /state/keyvalues/word-count/all often
+        # responds with 500 error.
+
+        driver.stop()
+        example2.stop()
+        example1.stop()
+
+
+class KafkaStreamsMicroservices(RedpandaTest):
+    """
+    Test KafkaStreams Microservices examples which demonstrate an order management
+    workflow, such as one might find in retail and online shopping.
+    """
+
+    topics = (
+        TopicSpec(name="orders"),
+        TopicSpec(name="warehouse-inventory"),
+        TopicSpec(name="order-validations"),
+        TopicSpec(name="payments"),
+        TopicSpec(name="customers"),
+        TopicSpec(name="platinum"),
+        TopicSpec(name="gold"),
+        TopicSpec(name="silver"),
+        TopicSpec(name="bronze"),
+    )
+
+    def __init__(self, test_context):
+        super(KafkaStreamsMicroservices,
+              self).__init__(test_context=test_context)
+        self._ctx = test_context
+
+    @cluster(num_nodes=4)
+    @matrix(jar=[
+        "EmailServiceTest", "FraudServiceTest", "InventoryServiceTest",
+        "OrderDetailsServiceTest", "OrdersServiceTest",
+        "ValidationsAggregatorServiceTest"
+    ])
+    def test_kafka_streams_microservices(self, jar):
+        example_jar = KafkaStreamExamples.MicroservicesExample(
+            self.redpanda, jar)
+        example = ExampleRunner(self._ctx, example_jar, timeout_sec=TIMEOUT)
+        example.start()
+        example.wait()
+        example.stop()
+
+
+# Not using any of the KafkaStreams base classes because this test
+# runs a client-side script to "reset" the app in the middle of execution
+# that other tests do not use.
+class KafkaStreamsAppReset(RedpandaTest):
+    """
+    Test KafkaStreams ApplicationReset example that uses confluent's cleanUp API
+    to reset an application between runs.
+    """
+    topics = (
+        TopicSpec(name="my-input-topic"),
+        TopicSpec(name="my-output-topic"),
+        TopicSpec(name="rekeyed-topic"),
+    )
+
+    def __init__(self, test_context):
+        super(KafkaStreamsAppReset, self).__init__(test_context=test_context)
+        self._ctx = test_context
+
+    def is_valid_msg(self, msg):
+        key = msg["key"]
+        return "record-1" in key or "record-2" in key or "record-3" in key
+
+    @cluster(num_nodes=6)
+    def test_kafka_streams_app_reset(self):
+        example_jar = KafkaStreamExamples.AppResetExample(self.redpanda)
+        example = ExampleRunner(self._ctx, example_jar, timeout_sec=TIMEOUT)
+        producer = AnomalyProducer(self._ctx, self.redpanda,
+                                   self.topics[0].name)
+        consumer = RpkConsumer(self._ctx, self.redpanda, self.topics[1].name)
+
+        example.start()
+
+        producer.start()
+        producer.wait()
         producer.stop()
+
+        consumer.start()
+        consume(self._ctx, consumer, self.is_valid_msg)
```

Comment:
This line looks like we want to consume all messages from `self.topics[1]` and to check that all of them are valid. But if we look at the `consume` function it consumes until it reads the first valid message. It's worrisome that we check only until the first correct message.

Actually the whole validation looks weak. We only check that a key contains a substring:
  - we don't check order
  - we don't check value
  - we don't check cardinality of the output stream

I've commented out the `example.node.account.ssh(cmd)` and the test still passed. What are we trying to test with this test?

Reply:
The purpose of the AppResetExample is to demonstrate confluents `cleanUp` API with less focus on what the program is actually doing. See https://github.com/vectorizedio/kafka-streams-examples/blob/93baddb468a09ea3fb57fd1431e58e5d2ba24f3a/src/main/java/io/confluent/examples/streams/ApplicationResetExample.java#L34

> I've commented out the example.node.account.ssh(cmd) and the test still passed. What are we trying to test with this test?

Confluents `cleanUp` API is on the client-side. The cleanUp is activated with confluent's `kafka-streams-application-reset` shell script (generated via `KStreamsExamples.reset_app_cmd`).  The way this example works is:
1. Do some simple prod-cons such as WordCount
2. Stop the example and "reset" the app with the `cleanUp` API <-- this is invoked with `example.node.account.ssh(cmd)`
3. Re-run the app and check that the output from the consumer is still valid.

Step 2 is what we're testing.

Reply:
> Actually the whole validation looks weak. We only check that a key contains a substring:
>    we don't check order
>    we don't check value
>    we don't check cardinality of the output stream

I'll make the tests stronger in my next push.